In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure
import statsmodels.graphics.tsaplots as sgt
import statsmodels.tsa.stattools as sts
import os
import statsmodels.api as sm
import statsmodels.tsa.stattools as st
import seaborn as sns
sns.set()

In [2]:
mydataraw = pd.read_csv("oilpriceindex.csv")

In [3]:
mydatats = pd.read_csv("oilpriceindex.csv", na_values=["null"], index_col="DATE", parse_dates=True, infer_datetime_format=True)

In [4]:
mydatats= mydatats.fillna(method="bfill")

In [5]:
mydatats.isna().sum()

WTI      0
BRENT    0
INR      0
NIFTY    0
dtype: int64

In [7]:
pip install arch

     |████████████████████████████████| 808kB 3.8MB/s 


In [8]:
from arch import  arch_model

In [10]:
mydatats["WTI_PCT"]= mydatats.WTI.pct_change(1)*100
mydatats["WTI_sqr"]= mydatats.WTI_PCT.mul(mydatats.WTI_PCT)

In [11]:
mydatats.head()

WTI  BRENT     INR  NIFTY    WTI_PCT     WTI_sqr
DATE                                                          
1990-01-01  22.86  21.25  16.880   0.44        NaN         NaN
1990-02-01  22.11  19.81  16.960   2.45  -3.280840   10.763910
1990-03-01  18.43  18.39  17.300  -0.28 -16.644052  277.024482
1990-04-01  18.43  16.61  17.300  -6.18   0.000000    0.000000
1990-05-01  18.20  16.35  17.378  -1.62  -1.247965    1.557417

In [18]:
model_arch1 = arch_model(mydatats.WTI_PCT[1:])
estimates_arch1 = model_arch1.fit(update_freq= 7)
estimates_arch1.summary()

Iteration:      7,   Func. Count:     48,   Neg. LLF: 1306.3797916933079
Optimization terminated successfully.    (Exit mode 0)
            Current function value: 1306.3797877475527
            Iterations: 8
            Function evaluations: 54
            Gradient evaluations: 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                     Constant Mean - GARCH Model Results                      
==============================================================================
Dep. Variable:                WTI_PCT   R-squared:                       0.000
Mean Model:             Constant Mean   Adj. R-squared:                  0.000
Vol Model:                      GARCH   Log-Likelihood:               -1306.38
Distribution:                  Normal   AIC:                           2620.76
Method:            Maximum Likelihood   BIC:                           2636.47
                                        No. Observations:                  375
Date:                Fri, Jul 02 2021   Df Residuals:                      374
Time:                        13:51:09   Df Model:                            1
                                 Mean Model                                
===========================================================================
                 coef    std err          t      P>|t|     95.0% Conf. Int.
---------------------------------------------------------------------------
mu             0.6505      0.382      1.703  8.865e-02 [-9.835e-02,  1.399]
                             Volatility Model                             
==========================================================================
                 coef    std err          t      P>|t|    95.0% Conf. Int.
--------------------------------------------------------------------------
omega          6.7086      4.326      1.551      0.121   [ -1.770, 15.187]
alpha[1]       0.1266  5.912e-02      2.142  3.223e-02 [1.074e-02,  0.243]
beta[1]        0.7669      0.110      6.965  3.281e-12   [  0.551,  0.983]
==========================================================================

Covariance estimator: robust
"""

**Simple ARCH(1)**

**Constant Mean**

In [19]:
model_arch1 = arch_model(mydatats.WTI_PCT[1:], mean= "constant", vol="ARCH", p = 1)
estimates_arch1 = model_arch1.fit(update_freq= 4)
estimates_arch1.summary()

Iteration:      4,   Func. Count:     24,   Neg. LLF: 1315.3454292119245
Iteration:      8,   Func. Count:     44,   Neg. LLF: 1315.2964907824723
Optimization terminated successfully.    (Exit mode 0)
            Current function value: 1315.2954492327399
            Iterations: 10
            Function evaluations: 54
            Gradient evaluations: 10


<class 'statsmodels.iolib.summary.Summary'>
"""
                      Constant Mean - ARCH Model Results                      
==============================================================================
Dep. Variable:                WTI_PCT   R-squared:                       0.000
Mean Model:             Constant Mean   Adj. R-squared:                  0.000
Vol Model:                       ARCH   Log-Likelihood:               -1315.30
Distribution:                  Normal   AIC:                           2636.59
Method:            Maximum Likelihood   BIC:                           2648.37
                                        No. Observations:                  375
Date:                Fri, Jul 02 2021   Df Residuals:                      374
Time:                        13:56:18   Df Model:                            1
                                Mean Model                                
==========================================================================
                 coef    std err          t      P>|t|    95.0% Conf. Int.
--------------------------------------------------------------------------
mu             0.8935      0.411      2.175  2.963e-02 [8.835e-02,  1.699]
                             Volatility Model                             
==========================================================================
                 coef    std err          t      P>|t|    95.0% Conf. Int.
--------------------------------------------------------------------------
omega         50.6598      5.443      9.308  1.306e-20   [ 39.992, 61.327]
alpha[1]       0.2780      0.122      2.274  2.295e-02 [3.842e-02,  0.518]
==========================================================================

Covariance estimator: robust
"""

Alpha = 0.2780 = Impact of past error square 



Beta =  Persistence (A +B <=1)

In [20]:
model_arch1 = arch_model(mydatats.WTI_PCT[1:], mean= "zero", vol="ARCH", p = 1)
estimates_arch1 = model_arch1.fit(update_freq= 4)
estimates_arch1.summary()

Iteration:      4,   Func. Count:     19,   Neg. LLF: 1317.5880973740636
Optimization terminated successfully.    (Exit mode 0)
            Current function value: 1317.5876495217628
            Iterations: 6
            Function evaluations: 27
            Gradient evaluations: 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                        Zero Mean - ARCH Model Results                        
==============================================================================
Dep. Variable:                WTI_PCT   R-squared:                       0.000
Mean Model:                 Zero Mean   Adj. R-squared:                  0.003
Vol Model:                       ARCH   Log-Likelihood:               -1317.59
Distribution:                  Normal   AIC:                           2639.18
Method:            Maximum Likelihood   BIC:                           2647.03
                                        No. Observations:                  375
Date:                Fri, Jul 02 2021   Df Residuals:                      375
Time:                        13:59:19   Df Model:                            0
                             Volatility Model                             
==========================================================================
                 coef    std err          t      P>|t|    95.0% Conf. Int.
--------------------------------------------------------------------------
omega         52.6224      5.406      9.734  2.152e-22   [ 42.027, 63.218]
alpha[1]       0.2481      0.119      2.080  3.751e-02 [1.433e-02,  0.482]
==========================================================================

Covariance estimator: robust
"""

In [21]:
model_garch_1_1 = arch_model(mydatats.WTI_PCT[1:], mean= "zero", vol="GARCH", p = 1, q = 1)
estimates_arch1 = model_garch_1_1.fit(update_freq= 4)
estimates_arch1.summary()

Iteration:      4,   Func. Count:     25,   Neg. LLF: 1307.8048202555992
Optimization terminated successfully.    (Exit mode 0)
            Current function value: 1307.7957893936746
            Iterations: 6
            Function evaluations: 36
            Gradient evaluations: 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                       Zero Mean - GARCH Model Results                        
==============================================================================
Dep. Variable:                WTI_PCT   R-squared:                       0.000
Mean Model:                 Zero Mean   Adj. R-squared:                  0.003
Vol Model:                      GARCH   Log-Likelihood:               -1307.80
Distribution:                  Normal   AIC:                           2621.59
Method:            Maximum Likelihood   BIC:                           2633.37
                                        No. Observations:                  375
Date:                Fri, Jul 02 2021   Df Residuals:                      375
Time:                        14:02:28   Df Model:                            0
                             Volatility Model                             
==========================================================================
                 coef    std err          t      P>|t|    95.0% Conf. Int.
--------------------------------------------------------------------------
omega          6.2162      3.862      1.609      0.108   [ -1.354, 13.786]
alpha[1]       0.1237  5.508e-02      2.246  2.471e-02 [1.575e-02,  0.232]
beta[1]        0.7780  9.827e-02      7.917  2.439e-15   [  0.585,  0.971]
==========================================================================

Covariance estimator: robust
"""

In [23]:
# alpha[1]	0.1237

# beta[1]	0.7780

Sensex = A = 2.5  B = 

NIKKE A= 5.1 B = 3.2